tfjs-backend-webgpu/src/lrn_webgpu.ts

```diff
@@ -0,0 +1,145 @@
+ * limitations under the License.
+ * =============================================================================
+ */
+
+import {util} from '@tensorflow/tfjs-core';
+import {getMainHeaderString as main, WebGPUProgram} from './webgpu_program';
+import {computeDispatch, flatDispatchLayout} from './webgpu_util';
+
+const powOperatorSnippet = `
+  var powValue = 0.0;
+  let basis = uniforms.bias + uniforms.alpha * sum;
+  if (uniforms.beta == 0.5) {
+    powValue = inverseSqrt(basis);
+  } else if (uniforms.beta == 1.0) {
+    powValue = 1.0 / basis;
+  } else {
+    powValue = exp(log(basis) * (-uniforms.beta));
+}
+`;
+
+export class LRNProgram implements WebGPUProgram {
+  outputShape: number[] = [];
+  shaderKey: string;
+  dispatchLayout: {x: number[]};
+  dispatch: [number, number, number];
+  variableNames = ['x'];
+  uniforms = 'radius : i32, bias : f32, alpha : f32, beta : f32,';
+  workgroupSize: [number, number, number] = [64, 1, 1];
+  size = true;
+
+  constructor(xShape: number[]) {
+    this.outputShape = xShape;
+    this.dispatchLayout = flatDispatchLayout(this.outputShape);
+    this.dispatch = computeDispatch(
+        this.dispatchLayout, this.outputShape, this.workgroupSize);
+    this.shaderKey = 'lrn';
+  }
+
+  getUserCode(): string {
+    const userCode = `
+    ${main('index')} {
+      if (index < uniforms.size) {
+        let coords = getOutputCoords();
+        let b = coords[0];
+        let r = coords[1];
+        let c = coords[2];
+        let d = coords[3];
+
+        let x = getX(b, r, c, d);
+        var sum = 0.0;
+        for (var i = -uniforms.radius; i <= uniforms.radius; i = i + 1) {
+          let idx = d + i;
+          if (idx >= 0 && idx < uniforms.xShape[3]) {
+            let z = getX(b, r, c, idx);
+            sum = sum + z * z;
+          }
+        }
+        ${powOperatorSnippet}
+
+        setOutputAtIndex(index, x * powValue);
+      }
+    }
+  `;
+    return userCode;
+  }
+}
+
+export class LRNSharedProgram implements WebGPUProgram {
+  outputShape: number[] = [];
+  shaderKey: string;
+  dispatchLayout: {x: number[], y: number[], z: number[]};
+  dispatch: [number, number, number];
+  variableNames = ['x'];
+  uniforms = 'radius : i32, bias : f32, alpha : f32, beta : f32,';
+  workgroupSize: [number, number, number] = [256, 1, 1];
+  maxAllowRadius = 16;
+  elementsPerWorkgroup: number;
+
+  constructor(xShape: number[], radius: number) {
+    util.assert(
+        radius <= this.maxAllowRadius,
+        () => `Radius must be less than or equal to ${
+            this.maxAllowRadius}, current radius is ${radius}`);
+
+    this.outputShape = xShape;
+    this.elementsPerWorkgroup = this.workgroupSize[0] - 2 * radius;
+    this.dispatchLayout = {x: [3], y: [2], z: [0, 1]};
+    this.dispatch = computeDispatch(this.dispatchLayout, this.outputShape, [
+      this.elementsPerWorkgroup, this.workgroupSize[1], this.workgroupSize[2]
```

Comment:
Can we directly use the `workgroupSize` here and allocate workgroup memory `lrn_Sub` as `this.workgroupSize[0] + 2 * radius`? Will it be more easier to understand? Meanwhile, `radium` should be as part of the `shaderKey`.

Reply:
Any why there is `maxAllowRadius = 16` limitation?

Reply:
Jiajia, I had added two comments to explain the reasons. And the fix the error from `this.workgroupSize[0] + 2 * radius` to `this.workgroupSize[0] - 2 * this.maxAllowRadius`.
Is it OK?